feat: RB -> OS lifecycle

crepererum opened the pull request:
Closes #1083.

alamb commented on data_types/src/database_rules.rs:
```diff
@@ -181,7 +181,7 @@ pub struct LifecycleRules {
 
     /// Once the total amount of buffered data in memory reaches this size start
-    /// dropping data from memory based on the drop_order
+    /// dropping data from memory based on the [`sort_order`](Self::sort_order)
```
I *think* rustdoc is smart enough to figure out the link target in this case. Something like

```suggestion
    /// dropping data from memory based on the [`sort_order`]
```

crepererum replied:
I think that doesn't work with fields, just with types:

## ``[`sort_order`](Self::sort_order)``
![Screenshot_2021-04-13 data_types database_rules LifecycleRules - Rust2](https://user-images.githubusercontent.com/1529400/114512677-a7f74080-9c39-11eb-8e31-6c5e5532dee6.png)

## ``[`sort_order`]`` / `[sort_order]`
![Screenshot_2021-04-13 data_types database_rules LifecycleRules - Rust](https://user-images.githubusercontent.com/1529400/114513018-06242380-9c3a-11eb-95f4-62be8cf810a7.png)

alamb commented on data_types/src/database_rules.rs:
```diff
@@ -199,6 +199,9 @@ pub struct LifecycleRules {
     /// Allow dropping data that has not been persisted to object storage
     pub drop_non_persisted: bool,
 
+    /// Allow chunks to be persisted to object storage.
```
Does this setting force / ensure that chunks are persisted to object storage before drop? If so, perhaps we should remove the `Allow` which makes it sound to me as if if persisting chunks is optional.

```suggestion
    /// Persists chunks to object storage.
```

alamb commented on generated_types/protos/influxdata/iox/management/v1/database_rules.proto:
```diff
@@ -151,6 +151,9 @@ message LifecycleRules {
   // Allow dropping data that has not been persisted to object storage
   bool drop_non_persisted = 7;
 
+  // Allow chunks to be persisted to object storage.
```
same suggestion as the comment on `database_rules.rs`

alamb commented on server/src/db.rs:
```diff
@@ -558,6 +559,41 @@ impl Db {
 
+    /// Spawns a task to perform
+    /// [`load_chunk_to_object_store`](Self::load_chunk_to_object_store)
+    pub fn load_chunk_to_object_store_in_background(
```
I wonder if this might be more consistently named  `write_chunk_to_object_store_in_background` (consistent with `Job::WriteChunk`)?

```suggestion
    pub fn write_chunk_to_object_store_in_background(
```

(though I realize you are following the pattern of `load_chunk_to_object_store`)

crepererum replied:
Let me do that last, AFTER resolving all other issues and test questions to avoid rebasing back and forth.

alamb commented on server/src/db/lifecycle.rs:
```diff
@@ -489,16 +550,31 @@ mod tests {
         chunks[2].set_closing().unwrap();
         chunks[2].set_moving().unwrap();
         chunks[2].set_moved(Arc::clone(&rb)).unwrap();
+        chunks[3].set_closing().unwrap();
```
Could you perhaps add some comments to this test about what it is testing? It seems like it is modeling two chunks that are written to the object store via manually triggered lifecycle events and then ensuring that the mover doesn't move them?

crepererum replied:
Added comments and pulled out the "transition to state X" boilerplate. Should be easier to understand now.

alamb commented on server/src/db/lifecycle.rs:
```diff
@@ -489,16 +550,31 @@ mod tests {
-        assert_eq!(mover.events, vec![MoverEvents::Drop(2)]);
+        assert_eq!(
+            mover.events,
+            vec![MoverEvents::Drop(2), MoverEvents::Drop(4)]
```
I was confused at first that there were no` MoverEvents::Write` events here but it is because the `rules` don't have `persist`, is that right?

crepererum replied:
Yes. I've added a comment at the top of the test to point out that fact. Let me know if that helps.

alamb replied:
it is clearer now -- thank you

alamb commented on server/src/db/lifecycle.rs:
```diff
@@ -512,4 +588,41 @@ mod tests {
         mover.check_for_work(from_secs(10));
         assert_eq!(mover.events, vec![]);
     }
+
```
As a style preference, I would personally suggest adding a `test_no_persist()` here that tested the scenarios that chunks aren't persisted if the life cycle rules have `persist=false` rather than adding on to the `test_buffer_size_soft()` test 

The rationale for doing so would be to make it clear what each test was testing (it may also avoid a need for additional comments)

NGA-TRAN replied:
+1

crepererum replied:
I think you're confusing two things here:

1. `persist=false` leads to not persisting things. For that I can add an explicit test.
2. The `test_buffer_size_soft` however does another thing which is not (directly) linked to that flag: dropping ONLY "moved" and "written" chunks. ALL other states MUST be kept. And this is what `test_buffer_size_soft` also ensures.

Or in other words: this PR makes two changes to the lifecycle core: the (optional, if `persist=true` transition from "moved" to "writing"; and the *fix* that the already implement drop section does not drop "writing" (aka in-flight) chunks.

alamb replied:
Good points -- I think making it clear the tests are validating "chunk mover only drops moved and written chunks" (either via comments, or a new test with a clear name) would improve the readability of the code. Thank you for the explination

alamb reviewed (commented):
I had some stylistic suggestions, but really the only thing I see missing from this PR is a test for  `load_chunk_to_read_buffer_in_background`

Really nice work @crepererum 👍

NGA-TRAN commented on server/src/db/lifecycle.rs:
```diff
@@ -512,4 +588,41 @@ mod tests {
+            new_chunk(1, Some(0), Some(0)),
+            new_chunk(2, Some(0), Some(0)),
+        ];
+        chunks[0].set_closing().unwrap();
```
Adding comments at every `set` here will help reviewers and readers follow what each set means and actions triggered by  them

crepererum replied:
done, see https://github.com/influxdata/influxdb_iox/pull/1174#discussion_r612217951

NGA-TRAN commented:
@crepererum It is amazing you have this PR out so quickly.
I have a a few minor detailed comments. My major comment here is in some of your set_* actions that trigger writing chunks to object store, can you capture the output data of the object stores to make sure they actually happen? These [one_chunk_one_table](https://github.com/influxdata/influxdb_iox/blob/a728494c760c168be0f19b2bc89e68eb9b0913fc/server/src/db.rs#L936)   and [one_chunk_many_tables](https://github.com/influxdata/influxdb_iox/blob/a728494c760c168be0f19b2bc89e68eb9b0913fc/server/src/db.rs#L1024) have some examples how to read data from parquet files. I did not see the input data of the tests, either. Maybe that is something missing in @alamb 's comment "but really the only thing I see missing from this PR is a test for load_chunk_to_read_buffer_in_background"

crepererum commented on generated_types/protos/influxdata/iox/management/v1/database_rules.proto:
```diff
@@ -151,6 +151,9 @@ message LifecycleRules {
   bool drop_non_persisted = 7;
 
+  // Allow chunks to be persisted to object storage.
+  bool persist = 9;
```
As per [this internal conversion](https://influxdata.slack.com/archives/CRK9M8L5Q/p1618299148298500), these fields are subject to backwards compatibility and therefore we currently do NOT renumber existing fields. We might do a complete re-enumeration before cutting a proper release.

crepererum commented:
@alamb I've added a test that you've requested. Let me know if that's what you've had in mind. I've also addressed all other comments (except for the rename, will do that last).

tustvold commented on server/src/db/lifecycle.rs:
```diff
@@ -129,8 +149,9 @@ trait ChunkMover {
-                        if rules.drop_non_persisted
-                            || matches!(chunk_guard.state(), ChunkState::Moved(_))
+                        if (rules.drop_non_persisted
+                            || matches!(chunk_guard.state(), ChunkState::Moved(_)))
```
I think the `ChunkState::Moved(_)` should just be removed now? This was just a hack whilst the persistence story was being wired up

crepererum replied:
@tustvold I don't think so. Imagine a case where we are able to get LOADS of incoming data w/o being able to write them off the OS fast enough. In that case, dropping unpersisted chunks is some kind of "last resort". That's also why we have two flags:

- `persist`: are we persisting at all?
- `drop_non_persisted`: if we don't persist, this is required to drop any data all. If we persist, then this acts as some kind of "overflow protection"

tustvold replied:
My expectation would be that drop_non_persisted would never drop unpersisted data, and the server would instead start rejecting writes once it hit the hard buffer size limit (which isn't currently implemented)?

crepererum replied:
Let me re-phrase my answer above, because I think we're probably talking about the same thing:

`persist` | `drop_non_persisted` | what happens
 -- | -- | --
`false` | `false` | No data is ever dropped. When size limit is reached writes are rejected (not sure if the latter one is currently implemented).
`false` | `true` | Data is never written to OS. When size limit is reached data is just dropped.
`true` | `false` | Data is written to OS. When size limit is reached persisted data is dropped. If we cannot persist fast enough, writes are rejected (not sure if the latter one is currently implemented).
`true` | `true` | Data is written to OS. When size limit is reached persisted data is dropped. If we cannot persist fast enough unpersisted data will be dropped as well.

Does that makes sense?

crepererum replied:
Oh, good catch, the `||` should be a `&&`. Will fix and adjust the test.

tustvold reviewed (approved):
This looks good to me, very nice :+1:

NGA-TRAN reviewed (approved):
It looks great, @crepererum

alamb commented on server/src/db.rs:
```diff
@@ -1671,4 +1707,41 @@ mod tests {
+            );
+        }
+
+        // we should have chunks in the mutable buffer, read buffer, and object store
```
👍
